File: backend/app/services/request_builder/mapping.py
"""
字段映射（Mapping）—— 将内部字段改名/移动到嵌套路径。

mapping 格式：
    {"source_field": "target.dot.path"}  → 移动字段到嵌套路径
    {"source_field": "arr[].nested"}     → 数组展开：每个元素包装为 {"nested": element}
    {"source_field": "arr[]"}            → 数组直接赋值（不包装）
    {"field_to_remove": None}            → 删除字段
"""

from __future__ import annotations

from typing import Any


def apply_mapping(body: dict, mapping: dict[str, str | None]) -> dict:
    """按 mapping 配置重映射/删除字段（原地修改）。

    Args:
        body: 当前请求体
        mapping: 字段映射表，值可以是目标路径字符串或 None（删除）

    Returns:
        修改后的 body（同一引用）
    """
    if not mapping:
        return body
    for src, target in mapping.items():
        if target is None:
            # 删除字段
            body.pop(src, None)
        elif isinstance(target, str) and src in body:
            value = body.pop(src)
            _set_nested(body, target, value)
    return body


def _set_nested(d: dict, path: str, value: Any) -> None:
    """将 value 设置到嵌套路径，支持 [] 数组展开语法。

    示例：
        path='extra_body.image'                   → d['extra_body']['image'] = value
        path='images[].image_url', value=['u1','u2'] → d['images'] = [{'image_url':'u1'}, {'image_url':'u2'}]
        path='images[]',           value=['u1','u2'] → d['images'] = ['u1','u2']
    """
    # ---------- 数组展开路径 ----------
    if "[]" in path:
        if not isinstance(value, list):
            # 值不是列表，去掉 [] 按普通路径处理
            path = path.replace("[]", "")
            _set_nested(d, path, value)
            return

        prefix, rest = path.split("[]", 1)
        suffix = rest.lstrip(".")  # 去掉开头的 .

        # 构建前缀路径（不含 [] 和后续部分）
        if prefix.rstrip("."):
            prefix = prefix.rstrip(".")

        if not suffix:
            # "images[]" → 直接赋值列表
            _set_nested_raw(d, prefix, value)
            return

        # "images[].image_url" → 每个元素包装为 {"image_url": item}
        result = []
        for item in value:
            obj = {}
            _set_nested_raw(obj, suffix, item)
            result.append(obj)

        _set_nested_raw(d, prefix, result)
        return

    # ---------- 普通嵌套路径 ----------
    _set_nested_raw(d, path, value)


def _set_nested_raw(d: dict, path: str, value: Any) -> None:
    """不处理 [] 语法的纯嵌套路径写入。"""
    if not path:
        # 空路径（前缀路径为空时可能出现）
        return
    parts = path.split(".")
    for part in parts[:-1]:
        if part not in d or not isinstance(d[part], dict):
            d[part] = {}
        d = d[part]
    d[parts[-1]] = value

File: server/services/request-builder/mapping.ts
/**
 * 字段映射引擎。
 * 提供嵌套路径移动、数组赋值、数组展开与字段删除四种映射语法。
 */

/**
 * 通用映射引擎，支持四种语法：
 *   - "source": "target.dot.path" — 嵌套路径移动
 *   - "source": "arr[]" — 数组直接赋值（key 保持原名）
 *   - "source": "xxx[].nested" — 数组展开（每个元素包装为 {nested: elem}，路径中 xxx 为任意数组名）
 *   - "source": null — 删除字段
 *
 * @param body 输入参数
 * @param mapping 映射规则表（{ fromKey: toPath | null }）
 */
export function applyMapping(
  body: Record<string, unknown>,
  mapping: Record<string, unknown> | null | undefined
): Record<string, unknown> {
  if (!mapping) return body;

  const result: Record<string, unknown> = {};
  const moved = new Set<string>();

  for (const [fromKey, toSpec] of Object.entries(mapping)) {
    // null → 删除
    if (toSpec === null || toSpec === undefined) {
      moved.add(fromKey);
      continue;
    }

    const toPath = String(toSpec);
    const value = body[fromKey];

    if (value === undefined) {
      moved.add(fromKey);
      continue;
    }

    // arr[] — 数组直接赋值（如 "ref_images": "arr[]" 表示保持原名，直接赋值数组）
    if (toPath === "arr[]") {
      result[fromKey] = value;
      moved.add(fromKey);
      continue;
    }

    // xxx[].nested — 数组展开（如 "ref_images": "images[].image_url"）
    // 匹配 "xxx[].yyy" 模式
    const arrExpandMatch = toPath.match(/^(\w+)\[\]\.(.+)$/);
    if (arrExpandMatch) {
      const arrName = arrExpandMatch[1]; // "images"
      const nestedPath = arrExpandMatch[2]; // "image_url"
      if (Array.isArray(value)) {
        setNested(result, `${arrName}`, value.map((item) => ({ [nestedPath]: item })));
      }
      moved.add(fromKey);
      continue;
    }

    // target.dot.path — 嵌套路径移动
    setNested(result, toPath, value);
    moved.add(fromKey);
  }

  // 保留未映射的字段
  for (const [key, val] of Object.entries(body)) {
    if (!moved.has(key)) {
      result[key] = val;
    }
  }

  return result;
}

function setNested(d: Record<string, unknown>, path: string, value: unknown): void {
  const parts = path.split(".");
  let current: Record<string, unknown> = d;
  for (let i = 0; i < parts.length - 1; i++) {
    const part = parts[i];
    if (!current[part] || typeof current[part] !== "object") {
      current[part] = {};
    }
    current = current[part] as Record<string, unknown>;
  }
  current[parts[parts.length - 1]] = value;
}
